### PR TITLE
Fix index out of range panic for kubectl alpha debug

### LINKED ISSUE #94579: kubectl alpha debug fails with "panic: runtime error: index out of range"
**What happened**:
When using `kubectl alpha debug` to create a debug container for a pod that has multiple init containers or ephemeral containers, `kubectl alpha debug` fails with `panic: runtime error: index out of range`.
```
# kubectl alpha debug myapp-pod --image=busybox
panic: runtime error: index out of range [1] with length 1

goroutine 1 [running]:
k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/debug.containerNameToRef(0xc00010a400, 0x18)
        /home/kiyoshim/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/debug/debug.go:547 +0x299
k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/debug.(*DebugOptions).computeDebugContainerName(0xc0000d8fc0, 0xc00010a400, 0x20, 0x19aa0a0)
        /home/kiyoshim/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/debug/debug.go:528 +0x4b
k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/debug.(*DebugOptions).generateDebugContainer(0xc0000d8fc0, 0xc00010a400, 0x2)
        /home/kiyoshim/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/debug/debug.go:389 +0x39
k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/debug.(*DebugOptions).debugByEphemeralContainer(0xc0000d8fc0, 0x1e93060, 0xc0000be010, 0xc00010a400, 0x1ed9a37800000000, 0xc000a079b8, 0xc00062a000, 0xc000650718, 0x4)
        /home/kiyoshim/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/debug/debug.go:359 +0x386
k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/debug.(*DebugOptions).visitPod(0xc0000d8fc0, 0x1e93060, 0xc0000be010, 0xc00010a400, 0x0, 0x0, 0x0, 0x1000100, 0x0)
        /home/kiyoshim/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/debug/debug.go:343 +0xc5
k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/debug.(*DebugOptions).Run.func1(0xc0007060e0, 0x0, 0x0, 0x0, 0x0)
        /home/kiyoshim/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/debug/debug.go:284 +0xd7
```

**What you expected to happen**:
`kubectl alpha debug` should create a debug container in pod `myapp-pod` successfully.

**How to reproduce it (as minimally and precisely as possible)**:
The EphemeralContainers feature gate must be enabled by using `--feature-gates=EphemeralContainers=true` flag in your cluster.
```
# cat myapp-pod.yaml
apiVersion: v1
kind: Pod
metadata:
  name: myapp-pod
spec:
  containers:
  - name: myapp-container
    image: busybox
    command: ['sh', '-c', 'echo The app is running! && sleep 1d']
  initContainers:
  - name: init-container-1
    image: busybox
    command: ['sh', '-c', 'sleep 1d']
  - name: init-container-2
    image: busybox
    command: ['sh', '-c', 'sleep 1d']

# kubectl apply -f myapp-pod.yaml
# kubectl alpha debug myapp-pod --image=busybox
panic: runtime error: index out of range [1] with length 1
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
```
# kubectl version 
Client Version: version.Info{Major:"1", Minor:"20+", GitVersion:"v1.20.0-alpha.0.996+74b04768b6d1a0", GitCommit:"74b04768b6d1a087c106828f45aeed2f84fa971e", GitTreeState:"clean", BuildDate:"2020-09-06T15:37:52Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"20+", GitVersion:"v1.20.0-alpha.0.996+74b04768b6d1a0", GitCommit:"74b04768b6d1a087c106828f45aeed2f84fa971e", GitTreeState:"clean", BuildDate:"2020-09-06T15:37:52Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
```

- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
NAME="CentOS Linux"
VERSION="7 (Core)"
- Kernel (e.g. `uname -a`):
Linux node-1 3.10.0-957.el7.x86_64 #1 SMP Thu Nov 8 23:39:32 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

/sig cli

### PR DESCRIPTION
**Pull request by kiyoshim55**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
When using `kubectl alpha debug` to create a debug container for a pod that has multiple init containers or ephemeral containers, `kubectl alpha debug` fails with `panic: runtime error: index out of range`. This PR fixes this issue.

**Which issue(s) this PR fixes**:
Fixes #94579

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
Fixed a panic in kubectl debug when pod has multiple init containers or ephemeral containers
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
None
```

/sig cli

### CODE REVIEW

**Comment by ialidzhikov** [comment]
Thank you @kiyoshim55!
Is it possible to add also an unit test about this case to prevent future regressions?

File: staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go
Comment (sulenn) [inline staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:548]: test the function

**Review by sulenn — APPROVED** [review]

File: staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go
Comment (sulenn) [inline staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go:550]: fghfghfgh

**Comment by kiyoshim55** [comment]
@ialidzhikov Thank you for the review! 
I added unit tests. `"pod with ephemeral containers"` test is added only for `GenerateDebugContainer` function because `GeneratePodCopyWithDebugContainer` function should not work with ephemeral containers. 

Please let me know if the changes are not what you expected. (e.g. Unit test should be added for `containerNameToRef` function itself.)

**Comment by verb** [comment]
@kiyoshim55 Thanks for the fix!

> I added unit tests. `"pod with ephemeral containers"` test is added only for `GenerateDebugContainer` function because `GeneratePodCopyWithDebugContainer` function should not work with ephemeral containers.

That's a good point, I think the current code will actually try to copy over the ephemeralContainers, though, which will cause the pod copy to be rejected by the API server. We should probably be dropping the ephemeralContainers from the copy.

It's only related to the bug you're fixing here, but do you feel like confirming this is the case, adding a test & fixing this bug, too?

**Comment by kiyoshim55** [comment]
@verb Thank you for your comment!
That's right. The current code will try to copy the ephemeralContainers and it will be rejected by the API server.

> It's only related to the bug you're fixing here, but do you feel like confirming this is the case, adding a test & fixing this bug, too?

Yes, I'm happy to do it. Would it be better to open another issue to address it because the original issue here is `"panic: runtime error: index out of range"` error?

**Comment by verb** [comment]
@kiyoshim55 great, thanks! You discovered it while adding tests for this issue, and it's a simple fix, so I recommend rolling them together, adding the complete set of tests for this issue and saving a bit of approver bandwidth.

**Comment by kiyoshim55** [comment]
@verb Thanks for your advice! I added a test and fixed the issue :) Please let me know if there are any points to be improved.

**Review by verb — APPROVED** [review]
Code LGTM, thanks! Please [squash](https://www.kubernetes.dev/docs/guide/pull-requests/#squashing) and add a [release note](https://www.kubernetes.dev/docs/guide/release-notes/).

Please note in the squashed commit message the ephemeral containers change which fixes a related bug discovered when adding tests.

**Comment by verb** [comment]
> /release-note-none

This should probably have a release note since it resolves a panic in kubectl.

**Review by seans3 — APPROVED** [review]
/lgtm

**Comment by seans3** [comment]
/hold

for release note

**Comment by kiyoshim55** [comment]
@verb @seans3 Thank you for your review! Commits squashed, and I added a release note.

**Comment by verb** [comment]
@soltysh @kiyoshim55 I think it's worth requesting this to be cherry picked into the release branch, what do you think?

**Comment by eddiezane** [comment]
@verb good call. I spoke with @seans3 yesterday and he agreed.

Just opened the cherry pick.

**Pull request merged**
